### ISSUE TITLE
Relative Links in Markdown to Anchors do not get translated correctly in HTML output

### ISSUE THREAD

**Issue by RobertClassen** [doxygen/doxygen#9701]
**Steps to reproduce:**
1. Download and unpack example: [DoxygenExample.zip](https://github.com/doxygen/doxygen/files/10098544/DoxygenExample.zip)
2. Open contained `Documentation~/index.html` file (static shortcut to generated `Documentation~/html/index.html`) and unfold hierarchy on the left, result should look like this: 
![DoxygenExample](https://user-images.githubusercontent.com/61211020/204141839-19339e72-6d9f-4c84-b461-27c63b46a229.png)
3. Mouse over and compare links for "SomeCategory" and "SomeOtherCategory" in left hierarchy vs right table → same links, works as expected
4. Compare links to individual methods → links in hierarchy work, links in table lead to "Page not found" notification in browser
5. (Open top-level `README.md` file in text editor and compare links at bottom)

**Expected result:** 
Appending the same relative link with `#` followed by the anchor (header) name should translate the generated link to the link in the hierarchy.
(Clicking a method name in the table should result in jumping/scrolling to and briefly highlighting the method name title the same way clicking the method name in the hierarchy does.)

**Use case:** 
The top-level `README.md` is used to give an overview of the contents on a GitHub repository page where the links in the table work.
I included `../README.md` in `INPUT` in the `Documentation~/Doxyfile` to show the same overview after opening the above `index.html` instead of showing an empty page and to allow for easier navigation via the custom table instead of having to either unfold and dig through the hierarchy or use the search function (which would require already knowing some class/method names to search for).

I tried searching but the closest issue I found was [this](https://stackoverflow.com/questions/59685012/doxygen-markdown-links-to-main-page-do-not-work), leading to https://github.com/doxygen/doxygen/pull/7511.

**Doxygen Version:** 
Tested with 1.9.1 and 1.9.5

**P.S.:** Possibly the same or a related issue exists with the "Contents" link under the "MyExample" title which links to the wrong `README.md` file, although this is just added as an extra example for relative links in the same `README.md` file while links to other `README.md` files are my main concern.

**Comment by albert-github** [comment]
Looking at 2 links represented in the table:
```
[SomeOtherCategory]: Runtime/Scripts/SomeOtherCategory/README.md
[SomeOtherCategory.MyOtherMethod0]: Runtime/Scripts/SomeOtherCategory/README.md#MyOtherMethod0
```
the first one, as you wrote, properly redirects as it translates to md__doxygen_example__runtime__scripts__some_other_category__r_e_a_d_m_e.html but the second one doesn't correctly redirect as no real translation is done and the resulting link points to Runtime/Scripts/SomeCategory/README.md#MyMethod0

The first problem we run into is the fact that in the README.md file there is no anchor with the name MyMethod0 (GitHub does some proprietary translation which is nowhere described and also not part of the of a "Markdown standard" like https://daringfireball.net/projects/markdown/syntaxor https://spec.commonmark.org/). Doxygen has its own labeling strategy (see https://www.doxygen.nl/manual/markdown.html#md_header_id) though this is to no avail here on its own.
The second problem looks like the fact that as soon as an anchor is specified to jump to the `.md` file link is not translated anymore.
We see that the first link is translated by doxygen for further processing into:
```
@ref D:/speeltuin/issue_9701/DoxygenExample/Runtime/Scripts/SomeCategory/README.md "SomeCategory"
```
whilst the second link is translated into:
```
<a href="Runtime/Scripts/SomeOtherCategory/README.md#MyOtherMethod0" ><tt>MyOtherMethod0</tt></a>
```

This has to be investigated a bit further.

**Comment by albert-github** [comment]
Regarding the part I mentioned as "second problem" I've just pushed a proposed patch, pull request #9703

Regarding the "first" problem, there is no solution for the automatic conversion of the section name to a label (as done by GitHub), so the user has to add the extended doxygen syntax for these anchors.

**Comment by RobertClassen** [comment]
I'm not familar with how Doxygen does this internally (so please correct me if I'm too far off) but I'd guess that at some point when it reads in all the relevant source files specified in `INPUT`, it checks if they have a `{#labelid}` pattern specified somewhere, then parses the given `labelid` and adds it to a `Dictionary<string, string>` as key and (either on-the-fly or later) generates the translated link as corresponding value (e.g. `md__doxygen_example__runtime__scripts__some_other_category__r_e_a_d_m_e.html`) which will be applied as `<a href>` url when the use of a `@ref` keyword is found?

Would it then not be possible to not only look for the `{#labelid}` pattern at the beginning but also for `# Header`, `## Subheader` (etc.) patterns (and corresponding patterns where the title is underscored by `===` or `---` in the next line)?

So basically with this new commit which allows detecting the file before the `#` correctly, check **that** file (e.g. `DoxygenExample\Runtime\Scripts\SomeOtherCategory\README.md`) for headers matching the anchor names which can now be detected?

> # SomeOtherCategory
> 
> ## MyOtherMethod0
> Calculates Lorem Ipsum.
> ```csharp
> MyOtherMethod0();
> ```
> 
> ## MyOtherMethod1
> Calculates Lorem Ipsum.
> ```csharp
> MyOtherMethod1();
> ```
> 
> ## MyOtherMethod2
> Calculates Lorem Ipsum.
> ```csharp
> MyOtherMethod2();
> ```

The problem with adding the `{#labelid}` pattern explicitly (as currently required) is that these would show up when viewing the `README.md` file on GitHub and probably also break the relative links over there instead which is not ideal...

If this approach of detecting headers might cause problems with other file types (e.g. in source code files of languages where `#` is used for comments), would it be possible to only do it this way for `.md` files?

**Comment by albert-github** [comment]
There are a number of flaws in the GitHub definition of the labels (to mention a few):
- there is no definition of how the labels are constructed, at least I've not found any reference to it
- the labels are not unique when having multiple (sub)headers with the same name, so a jump can only go to the first definition (at least in the documents I saw this was the case)
- doxygen labels are global, so not per file, and are thus allowed only once whilst the GitHub labels are local i.e. different pages can have the same label names

when at least the first problem is solved it is probably possible to think about solutions.

It is true that the doxygen way of labeling shows up on GitHUb, but the GitHub labeling strategy (which is also not standard) is unknown so we cannot even think about a conforming solution (although it would be nice to have a solution as it is a known problem).

**Comment by albert-github** [comment]
> there is no definition of how the labels are constructed, at least I've not found any reference to it

The main point is how how are header texts translated into labels, you already mentioned that non-alpha-numeric symbols were translated into `-` but there are also other characters (like: ☤ that lately I found in the pypa repository) that are completely removed (looks like non ASCII characters) so that is why a definition is needed.
The labels are not just needed for HTML but also for other output formats like LaTeX / PDF, RTF, Docbook, ...

> the labels are not unique when having multiple (sub)headers with the same name, so a jump can only go to the first definition (at least in the documents I saw this was the case)

This might not be important in your use case, but there are other usages like where one has:
```
Description method 1
-- Brief description
    ...
-- Detailed description
    ...
Description method 2
-- Brief description
    ...
-- Detailed description
   ...
```

and in that case a jump to "Brief Description" could go wrong.
I don't know what wikipedia uses for link generation (I should study the https://en.wikipedia.org/wiki/Help:Wikitex in that case), so I don't know how they handle this or if the prohibit it, but at first glance they use an other dedicated way of creating links.

> doxygen labels are global, so not per file, and are thus allowed only once whilst the GitHub labels are local i.e. different pages can have the same label names

Indeed here would be a big hassle as this would influence the complete system of links and anchors.
